### PR TITLE
crates/core: add try_params and more conversions

### PR DESCRIPTION
**Pull request by psarna**
This small PR adds a `try_params` macro, which works similarly to `params!`, except it returns a Result and is based on `TryInto` instead of `Into`.

I also evaluated how compatible with rusqlite we can/should be, and here are my conclusions:
1. rusqlite implements `Params` as a trait, while in our code it's a clonable struct
2. rusqlite params are based on a `ToSql` trait, as well as `&dyn ToSql` for passing parameters
3. our params are just Value instances passed by value
4. rusqlite does intricate lifetime management of the parameters, while we're more into "pass and clone everything as is to avoid being generic over lifetimes

To sum up, I *haven't* just migrated to rusqlite code in this PR, because rusqlite params are more in line with their lifetime management -> e.g. `query` returns a `Rows<'a>`, and the public interfaces are generally lifetime-bound. On the other hand, our params are just a structure, that can be copied over if necessary, which has a DX advantage - less fights with the borrow checker at the price of more memory copying and allocations.

Happy to discuss what should our final approach be!

### CODE REVIEW

**Comment by LucioFranco** [comment]
@psarna do you think we could support something similar to `&dyn ToSql`? I think that implementation is quite nice to support `&[..]` style api which imo is better than using macros.
